src/proxyp/Message.cc

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+#include "proxyp/Elements.h"
+#include "proxyp/Message.h"
+#include "sbuf/Stream.h"
+#include "sbuf/StringConvert.h"
+#include "SquidConfig.h"
+#include "StrList.h"
+
+ProxyProtocol::Message::Message(const SBuf &ver, const Two::Command cmd):
+    version_(ver),
+    command_(cmd),
+    ignoreAddresses_(false)
```

Comment:
Please do not. Some compilers (and some humans!) dislike the mixture of inlined and in-constructor data member initializations. We should use only the former (when we can) and only use the latter (otherwise).

Reply:
Lack of it makes several constructors undefined which results in move optimizations being forbidden.

Reply:
AFAICT the modern compilers define a bitmap from the static initializers that gets blitted into the memory space on newly allocated objects, that forms the default constructor. Other constructors run that same initialization before their more-specific construct logic (if any) to fill out the memory which remains undefined or needs to be overwritten.

Reply:
> Lack of it makes several constructors undefined which results in move optimizations being forbidden.

I am not sure why you mention this -- if doing X leads to compiler warnings, then any positive side effects of doing X are irrelevant in Squid context.

As for other constructors being left "undefined", it is good that they are automatically deleted/forbidden for this class because the class objects should not be copied (at least not without extra work that is better spent elsewhere).

> I have not found examples restricting their usage.

I suspect the compiler warnings in question are due to a compiler flaw/deficiency (i.e., it does not support something that it should support) and/or due to compiler authors belief that mixing initialization styles is either dangerous or disables something useful (i.e., it is just a compiler-specific warning, not a language-required prohibition). Unfortunately, I do not have a warning example to use for looking the warning explanation up in the compiler documentation.

It is also possible that my recollection about these warnings/restrictions is faulty. I believe Amos was the first who discovered this issue (in Squid context), but I do not have a ready-to-use reference to that email thread. I do not remember whether I have seen this warning myself, but it is likely that I have not.

@yadij, if this change request is no longer relevant, please mark this thread accordingly by clicking the "Resolve conversation" button. Otherwise, please post a conversation comment to clarify why the discussion so far has not addressed the change request.

Reply:
FWIW, looking at it in far more detail I found another members' sub-member type was not providing C++11 semantics yet either so this change alone would not be sufficient to enable the constructor optimizations anyway.

I still think we could do better by always using C++11 initialization - even when there are some members needing custom construct. But is not worth blocking progress at this point.